### DISCUSSION TITLE
WebSockets

### DISCUSSION THREAD

**Discussion by calvin2021y** [curl/curl#9107] [General]
Hi @bagder,  continue from https://github.com/curl/curl/pull/8995

I guess you mean use curl_ws_meta to the frame information like type is CURLWS_PONG, CURLWS_PING , CURLWS_TEXT ,CURLWS_BINARY, CURLWS_CLOSE ?

There is a MAX_WS_SIZE=4096 from curl_ws_send,  this limit is too small and websoket frame can be large like 64 KB. (or more)

And there is a output size_t *sent, which is the limit for max frame size ( you write "largest single fragment" from ws.h)?

Websocket frame can be a full packet, if you split it into multi frame could break some function.  

I need this function accept a memory block as frame, with extra prefix padding space can be used as frame header.  and a callback to let the caller know the memory is ready to be released (by the caller).  and it will send as a full frame without break into multi frames.

**Comment by bagder** [comment]
> I guess you mean use curl_ws_meta to the frame information like type is CURLWS_PONG, CURLWS_PING , CURLWS_TEXT ,CURLWS_BINARY, CURLWS_CLOSE ?

Yes!

When websocket data is delivered to the write callback `curl_ws_meta` can be used to ask libcurl what "opcode" that was used. By default however, pings will be auto-responded by libcurl itself so they will not be passed to the write callback.

It would also tell the callback if the FIN bit is set or not so that it knows if to expect more frames for the same message.

> There is a MAX_WS_SIZE=4096 from curl_ws_send, this limit is too small and websoket frame can be large like 64 KB. (or more)

Sure, a frame can be 2^63 bytes. I decided to start out simpler and set a fixed limit in how large frames the implementation supports. I'm open for feedback on how sensible that is and what frame sizes we **must** support to interop widely. Can we even have a maximum frame size?

If we are going to support really large frames, we need to "stream" them to the callback instead of passing full frames as the current API says we do.

> Websocket frame can be a full packet, if you split it into multi frame could break some function.

A full packet of what?

> I need this function accept a memory block as frame,

Sorry, which function is *this function* referring to... ?

>  with extra prefix padding space can be used as frame header. and a callback to let the caller know the memory is ready to be released (by the caller). and it will send as a full frame without break into multi frames.

The functions already works with and expects full frames. But if we are going to support 2^63 frames, then the APIs need to be adjusted so that the entire frame is not necessarily provided in a single go...

**Comment by calvin2021y** [comment]
Thanks for the nice explain, I dont know there is a FIN exists to know there is more frame for same message. In this case I guess curl_ws_meta should also keep the message size ?

please ignore my question about memory block and release part. (This idea is  for message memory payload manage by user, avoid copy when send)

I just want to know how to send a message bigger than MAX_WS_SIZE in event loop ?  I should call curl_ws_send again and again until there is no left buffer to send ? (for example call 16 times for a 64K message)

**Comment by bagder** [comment]
> I dont know there is a FIN exists to know there is more frame for same message. In this case I guess curl_ws_meta should also keep the message size ?

If there is no FIN for a frame, there's another one coming for the same message, yes.

> I just want to know how to send a message bigger than MAX_WS_SIZE in event loop ?

That will be impossible. **If** there is a max size, there will not be a way to do something larger than the max. What would max mean if it isn't the max? But again: I'm interested in feedback on what the largest frames to support should be as I don't have much websockets experience myself. Are people/sites/services actually using that large frames?

**Comment by calvin2021y** [comment]
The message payload size almost always > 4K,  I guess 64K is a good limit.

for example people run MQTT by websocket, or some kind event stream. I can not think of a case size always < 4k.

**Reply by bagder** [reply]
Ok thanks! Let's bump the max to 64K for now then and see how people feel about that.

**Comment by bagder** [comment]
If anyone has a use case for websockets support and wants to work with me to try out the new websockets API early, I'm very interested in getting in touch.

I imagine that it will be beneficial to get something real tested out early on to see that the interface works out for something real.

Anyone?

**Comment by calvin2021y** [comment]
I would like to try use it. (current I am still libwebsockets)

I will try provide report here and let you know my problem with curl.

**Comment by bagder** [comment]
The initial WebSockets support was merged on September 9, 2022. run `./configure --enable-websockets` to get it added to your build. Still no particular adjustment for the command line tool for it.

**Comment by n0name321** [comment]
> Sure, a frame can be 2^63 bytes. I decided to start out simpler and set a fixed limit in how large frames the implementation supports. I'm open for feedback on how sensible that is and what frame sizes we must support to interop widely. Can we even have a maximum frame size?
If we are going to support really large frames, we need to "stream" them to the callback instead of passing full frames as the current API says we do.

The 64K max is still too small for some applications - servers which can be queried for larger chunks of data per request are not uncommon, and it may not be possible to know how big the response will be until it arrives (e.g. get all rows within a timespan).

There's another inconvenience in the current implementation: https://github.com/curl/curl/blob/7740530cedadcd85bace6d6fa93242d329a60eae/lib/ws.c#L410 The client is required to pass in a full sized buffer without knowing the length of the frame, so in practical terms this means the client must always pass in a 64K buffer (or do own parsing of the frame header, defeating the point of the ws library) or risk receiving an opaque CURLE_RECV_ERROR which is also used for genuine protocol/socket errors.

I believe the ability to receive >64K is necessary for wide adoption of curl for websockets, and adding streaming wouldn't even require a change to the current API: curl_ws_recv could fill up the passed in buffer up to buflen if the current frame is larger and return CURLE_AGAIN (or similar) to let the client know that they should keep polling for more data. This would solve both libcurl needing to keep arbitrarily-large internal frame buffers and users needing to know in advance how large a buffer to pass to libcurl

**Reply by bagder** [reply]
> The 64K max is still too small for some applications

I was afraid that would be the case. Thanks for this feedback. We need to consider how the API should behave so that we can work with (much) larger frames without them being painful or complicated for the application.

Ideally of course we would have someone with a use case that works against such a server, but I figure we might not be that lucky.

> I believe the ability to receive >64K is necessary for wide adoption of curl for websockets

Right, and to do that we need to address and change that "too small buffer" logic anyway so we might just as well take a bigger bite here and consider how to smoothly support "jumbo frames".

**Reply by n0name321** [reply]
I do have such a use case which is why I commented - I'm writing a client to connect to a crypto exchange Deribit, and for whatever reason the majority of such exchanges have decided to use websockets as their main protocol. For example this is the method you'd use to get a list of products tradeable for a particular currency: https://docs.deribit.com/#public-get_instruments , a pretty standard thing your client would need to do in order to know how to later send order requests to the server. 

Trying the current libcurl master, the connection works well to "ws://test.deribit.com/ws/api/v2" - HTTP to WS upgrade works, curl_ws_send `{   "method": "public/get_instruments", "params": { "currency": "BTC" }, "jsonrpc": "2.0", "id": 0}` is good but curl_ws_recv chokes on the ~280K reply and there's no option on the API for the server to paginate the results - you're getting 280K back today but next week the payload may be larger depending on what products happen to be tradeable.

If you'd like you can try the public API in a web interface at https://test.deribit.com/api_console, e.g. selecting method "public/get_instruments" & currency "BTC", click the WS button to perform websocket send + receive and see the messages to/from the server

**Reply by bagder** [reply]
Excellent. A rather big question for you then. If you were to update the current curl websocket API into something that would work for your case, what would you change and how?

**Comment by bagder** [comment]
I just submitted #9636 which changes the API for websocket to make it work with full size (63 bit) frames. I am very interested in feedback on the updated API proposed there.

**Comment by calvin2021y** [comment]
on my test I get this error:

```sh
* WS: send OPCODE BINARY
* WS: send FIN bit 1 (byte 82)
* WS: send payload len 1452
* OpenSSL SSL_write: SSL_ERROR_SYSCALL, errno 0
* WS: wanted to send 1460 bytes, sent 18446744073709551615 bytes
```

I use libuv with curl to send data async style,  call with curl_ws_send.

is the curl_ws_send exec block op here to rise this problem ?
